internal/gengapic/generator_test.go

```diff
@@ -0,0 +1,141 @@
+				Proto3Optional: proto.Bool(true),
+				Options:        optsUUID4,
+			},
+			{
+				Name:  proto.String("invalid_auto_populated_not_in_serviceconfig"),
+				Type:  typep(descriptor.FieldDescriptorProto_TYPE_STRING),
+				Label: labelp(descriptor.FieldDescriptorProto_LABEL_OPTIONAL),
+			},
+			{
+				Name:  proto.String("invalid_auto_populated_no_annotation"),
+				Type:  typep(descriptor.FieldDescriptorProto_TYPE_STRING),
+				Label: labelp(descriptor.FieldDescriptorProto_LABEL_OPTIONAL),
+			},
+			{
+				Name:    proto.String("invalid_auto_populated_required"),
+				Type:    typep(descriptor.FieldDescriptorProto_TYPE_STRING),
+				Label:   labelp(descriptor.FieldDescriptorProto_LABEL_REQUIRED),
+				Options: optsRequiredAndUUID4,
+			},
+			{
+				Name:    proto.String("invalid_auto_populated_int"),
+				Type:    typep(descriptor.FieldDescriptorProto_TYPE_INT64),
+				Label:   labelp(descriptor.FieldDescriptorProto_LABEL_OPTIONAL),
+				Options: optsUUID4,
+			},
+			{
+				Name:    proto.String("invalid_auto_populated_ipv4"),
+				Type:    typep(descriptor.FieldDescriptorProto_TYPE_STRING),
+				Label:   labelp(descriptor.FieldDescriptorProto_LABEL_OPTIONAL),
+				Options: optsIPV4,
+			},
+		},
+	}
+	outputType := &descriptor.DescriptorProto{
+		Name: proto.String("OutputType"),
+	}
+	file := &descriptor.FileDescriptorProto{
+		Package: proto.String("my.pkg"),
+		Options: &descriptor.FileOptions{
+			GoPackage: proto.String("mypackage"),
+		},
+	}
+	serv := &descriptor.ServiceDescriptorProto{
+		Name: proto.String("Foo"),
+	}
+
+	var g generator
+	g.opts = &options{
+		pkgName: "pkg",
+	}
+	g.imports = map[pbinfo.ImportSpec]bool{}
+
+	g.serviceConfig = &serviceconfig.Service{
+		Publishing: &annotations.Publishing{
+			MethodSettings: []*annotations.MethodSettings{
+				{
+					Selector: "my.pkg.Foo.GetOneThing",
+					AutoPopulatedFields: []string{
+						"request_id",
+						"invalid_auto_populated_no_annotation",
+						"invalid_auto_populated_required",
+						"invalid_auto_populated_int",
+						"invalid_auto_populated_ipv4",
+					},
+				},
+			},
+		},
+	}
+
+	commonTypes(&g)
+	g.descInfo.Type[".my.pkg."+inputType.GetName()] = inputType
+	g.descInfo.ParentFile[inputType] = file
+	g.descInfo.Type[".my.pkg."+outputType.GetName()] = outputType
+	g.descInfo.ParentFile[outputType] = file
+	g.descInfo.ParentFile[serv] = file
+
+	m := &descriptor.MethodDescriptorProto{
+		Name:       proto.String("GetOneThing"),
+		InputType:  proto.String(".my.pkg.InputType"),
+		OutputType: proto.String(".my.pkg.OutputType"),
+		Options:    &descriptor.MethodOptions{},
+	}
+	g.descInfo.ParentElement[m] = serv
+	serv.Method = []*descriptor.MethodDescriptorProto{m}
+
+	got := g.autoPopulatedFields(serv.GetName(), m)
+	if want := 1; len(got) != want {
+		t.Errorf("len(got) = %d, want: %d, got: %v", len(got), want, got)
+	}
+	if want := "request_id"; got[0].GetName() != want {
+		t.Errorf("got[0].GetName() = %s, want: %s", got[0].GetName(), want)
+	}
```

Comment:
This is not a required change. But another way to do this while being a little more future proof/exact is as follows:

```go
if diff := cmp.Diff(got, want, protocmp.Transform()); diff != "" {
  t.Errorf(...)
}
```

Where [`protocmp.Transform`](https://pkg.go.dev/google.golang.org/protobuf/testing/protocmp#Transform) is a protobuf provided `cmp.Option` comparing protobuf messages

Reply:
I tried using `cmp` first, but it turned out ugly due to unexported fields, so I switched to this.

Reply:
Did you use the option I mentioned above? It fixes issues with unexported comparisons.

Reply:
Fixed, thanks for the tip!